network: disable kernel creating prefix route when RouteTable= is set

ssahani opened the pull request:
Closes #8726.

yuwata commented:
If vrf is used, the prefix route created by kernel seems to have expected table id. So, Arch in KVM CI fails.

ssahani commented:
it's looking for the prefix route which no more there I guess. I dropped for DHCP.


```
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 2104, in test_dhcp_client_settings
    self.assertRegex(output, '192.168.5.0/24 via 192.168.5.5 proto dhcp')
AssertionError: Regex didn't match: '192.168.5.0/24 via 192.168.5.5 proto dhcp' not found in 'default via 192.168.5.1 proto dhcp src 192.168.5.181 metric 1024 \n192.168.5.1 proto dhcp scope link src 192.168.5.181 metric 1024'

======================================================================
FAIL: test_dhcp_client_vrf (__main__.NetworkdDHCPClientTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 2328, in test_dhcp_client_vrf
    self.assertRegex(output, 'inet 192.168.5.[0-9]*/24 brd 192.168.5.255 scope global dynamic veth99')
AssertionError: Regex didn't match: 'inet 192.168.5.[0-9]*/24 brd 192.168.5.255 scope global dynamic veth99' not found in '75: veth99@veth-peer: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master vrf99 state UP group default qlen 1000\n    link/ether 12:34:56:78:9a:bc brd ff:ff:ff:ff:ff:ff\n    inet 169.254.36.28/16 brd 169.254.255.255 scope link veth99\n       valid_lft forever preferred_lft forever\n    inet 192.168.5.181/24 brd 192.168.5.255 scope global dynamic noprefixroute veth99\n       valid_lft 3600sec preferred_lft 3600sec\n    inet6 2600::14/128 scope global dynamic noprefixroute \n       valid_lft 3595sec preferred_lft 3595sec\n    inet6 fe80::1034:56ff:fe78:9abc/64 scope link \n       valid_lft forever preferred_lft forever'

```

poettering commented:
Hmm, I don#t grok this, I'll leave this one to @yuwata to review/merge

ssahani commented:
The prefix routes are getting added to main table instead of the table mentioned in the config file.  ```IFA_F_NOPREFIXROUTE``` is precisely meant for this purpose to stop creating them. Instead of let them and create and delete .  Kernel looks at the flag if this is set then it does not create these routes which not not in the config file.

ssahani commented:
Here I added the creation logic done by @yuwata and added should be fine. @yuwata PTAL.

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -187,12 +187,37 @@ static int link_set_dhcp_routes(Link *link) {
 
+        /* Now add here the prefix routes */
+        table = link_get_dhcp_route_table(link);
+        if (table != RT_TABLE_MAIN) {
```
`table` may be set through VRF. In that case, `networkd` should not create the prefix route, and kernel creates it.

ssahani commented:
> It seems the basic logic is the same as my PR. What is the critical difference?

This PR is on steroids. It just the opposite and uses the flag. Like don't create a building and destroy it . Instead create it where it requires. This is also used by us (FLAG) theres. The unnecessary unreachable checks are not there . I would add the attribute logic to your name since it's the same code.

> also fallback logic for old kernels.

Too old kernel not sure who going to back-port such a newer version of Networkd. If we stick to that very difficult to maintain

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -188,12 +188,36 @@ static int link_set_dhcp_routes(Link *link) {
         /* Clear old entries in case the set was already allocated */
         set_clear(link->dhcp_routes);
 
-        table = link_get_dhcp_route_table(link);
```
Please do not drop this.

yuwata commented:
> It just the opposite and uses the flag.

You mean about `Address::flag`? My code and your code is equivalent. In both case the flag is set.
(BTW, naming of `Address::prefix_route` is misleading. When it is true, we add `IFA_F_NOPREFIXROUTE` flag... The value or naming should be inverted.)

> Like don't create a building and destroy it . Instead create it where it requires. 

Sorry, I cannot catch what you want to say. Could you elaborate more?

> The unnecessary unreachable checks are not there . 

What is the 'unreachable checks'? You mean the fallback logic I introduced in my PR?

> Too old kernel not sure who going to back-port such a newer version of Networkd. If we stick to that very difficult to maintain

CentOS CI fails without the fallback logic. Yeah, CentOS kernel is very old and is based on 3.11... But at least now, we can follow it.

ssahani commented:
> > It just the opposite and uses the flag.
> 
> You mean about `Address::flag`? My code and your code is equivalent. In both case the flag is set.
> (BTW, naming of `Address::prefix_route` is misleading. When it is true, we add `IFA_F_NOPREFIXROUTE` flag... The value or naming should be inverted.)
> 
> > Like don't create a building and destroy it . Instead create it where it requires.
> 
> Sorry, I cannot catch what you want to say. Could you elaborate more?
> 
> > The unnecessary unreachable checks are not there .
> 
> What is the 'unreachable checks'? You mean the fallback logic I introduced in my PR?
> 
> > Too old kernel not sure who going to back-port such a newer version of Networkd. If we stick to that very difficult to maintain
> 
> CentOS CI fails without the fallback logic. Yeah, CentOS kernel is very old and is based on 3.11... But at least now, we can follow it.



> > It just the opposite and uses the flag.
> 
> You mean about `Address::flag`? My code and your code is equivalent. In both case the flag is set.
> (BTW, naming of `Address::prefix_route` is misleading. When it is true, we add `IFA_F_NOPREFIXROUTE` flag... The value or naming should be inverted.)

It's accepted by everyone and kernel so there should be no conclusion.,
> 
> > Like don't create a building and destroy it . Instead create it where it requires.
> 
> Sorry, I cannot catch what you want to say. Could you elaborate more?

I can't understand what you are asking.
> 
> > The unnecessary unreachable checks are not there .
> 
> What is the 'unreachable checks'? You mean the fallback logic I introduced in my PR?
> 
> > Too old kernel not sure who going to back-port such a newer version of Networkd. If we stick to that very difficult to maintain
> 
> CentOS CI fails without the fallback logic. Yeah, CentOS kernel is very old and is based on 3.11... But at least now, we can follow it.

No We need to support to newer and old we just can skip .

yuwata commented:
Anyway, two points I want to ask are, 
- Why don't you delete the prefix route when the lease expires? You may already explained that, but unfortunately I cannot catch that.
- Do you want to say we should kill `CentOS CI (CentOS 7)` at least testing `networkd`? Even though there exist a fallback logic.

yuwata commented:
For audience of this thread. The basic approach of his and my PR are equivalent. But there exists two difference:
- In my PR, the prefix route is deleted when the lease expires.
- My PR also supports old kernels, but in that case, the prefix route is in main table.

ssahani commented:
> Anyway, two points I want to ask are,
> 
> * Why don't you delete the prefix route when the lease expires? You may already explained that, but unfortunately I cannot catch that.

Now kernel deletes automatically when addresses is gone as as already addd them back when lease gains (static)

> * Do you want to say we should kill `CentOS CI (CentOS 7)` at least testing `networkd`? Even though there exist a fallback logic.

What do I mean. Why you writing wrong things. I said older kernel does not support this newer version hence this test case can be skipped as we do for other.

ssahani commented:
> For audience of this thread. The basic approach of his and my PR are equivalent. But there exists two difference:
> 
> * In my PR, the prefix route is deleted when the lease expires.
> * My PR also supports old kernels, but in that case, the prefix route is in main table.

Where ? the prefix routes never gets created in the MAIN table at all it just skips and creates later while adding routes in the table mentioned in the configuration

ssahani commented:
But I have a better idea for what we do for the lentos CI and older kernel we test for the flag ```IFA_F_NOPREFIXROUTE``` . If it's not there then then we do adding or deleting else we do in newer way just to make happy the CI.

keszybz commented on src/network/networkd-dhcp4.c:
```diff
@@ -612,6 +636,11 @@ static int dhcp4_update_address(Link *link,
+        /* If DHCP route is set then stop creating prefix route. Instead add
+           it later on that table */
+        if (link->network->dhcp_route_table != RT_TABLE_MAIN)
+                addr->flags |= IFA_F_NOPREFIXROUTE;
```
This should also look at `link->network->dhcp_route_table_set`, see @yuwata's patch.

I merged the other PR, because it seems to be more carefully written (and CI passes).
